midend/flattenInterfaceStructs.cpp

```diff
@@ -57,21 +81,26 @@ const IR::StructInitializerExpression* StructTypeReplacement::explode(
         }
         vec->push_back(new IR::NamedExpression(f->name, expr));
     }
-    return new IR::StructInitializerExpression(fieldType->name, *vec, false);
+    return new IR::StructInitializerExpression(is_struct ? st->name : ht->name,
+                                               *vec, false);
 }
 
-static const IR::Type_Struct* isNestedStruct(const P4::TypeMap* typeMap, const IR::Type* type) {
+static const IR::Type* isNestedStruct(const P4::TypeMap* typeMap, const IR::Type* type) {
     if (auto st = type->to<IR::Type_Struct>()) {
         for (auto f : st->fields) {
             auto ft = typeMap->getType(f, true);
-            if (ft->is<IR::Type_Struct>())
+            if (ft->is<IR::Type_Struct>()) {
                 return st;
-        }
+            }
+            if (ft->is<IR::Type_Header>()) {
```

Comment:
Does this return the type `header hdr` if applied to `struct metadata`?
```
header hdr {
   bit<8> h0;
}
struct metadata {
   hdr hdr0;
}
```

Reply:
Yes.  I have checked in a new version of p4c/testdata/p4_16_sample/issue363-bmv2.p4.  In this P4 file you will see the following data structs which are same as yours.

`header bitvec_hdr {
    bit<5> g2;
    bit<3> g3;
    macDA_t mac;
}

struct parsed_packet_t {
    bitvec_hdr bvh;
};
`

Reply:
Does it handle the case where you may have
```
struct parsed_packet_t {
   bit<8> f0;
   bitvec_hdr bvh;
};
```
or
```
struct parsed_packet_t {
   bitvec_hdr bvh0;
   bitvec_hdr bvh1;
};
```

Reply:
Yes.  I have edited issue363-bmv2.p4 to included another bitvec_hr bvh1 like you said.  The reson is because isNestedStruct is called in a loop.  I will checkin the new issue363-bmv2.p4 shortly.  I do have to test a bit more by inspecting flattened IR, add few more tests.  Then I will move to another bmv2 midend pass.